### ISSUE TITLE
Catch Response of a "page endpoint" into a svelte page.

### ISSUE THREAD

**Issue by giuseppe-palmeri** [sveltejs/kit#4394]
### Describe the problem

I've been using SvelteKit for a while now to build a company site and a little PWA as a demo to show.

In this context, I feel the lack of both "layout page endpoints" (as has already been reported #4274) 
and the possibility of having control over te Response returned by a "page endpoint".

Currently the response from a "page endpoint" is set in props: Props (Record&lt;string, any&gt;) 
which is an argument passed to the load function if exported.

In case of error, however, the control is held by SvelteKit making sure to present an Internal Server Error (http status 500).

It would not be a problem under normal circumstances, but in some cases it would be very useful to be able to condition the program flow based on whether the call to the "page endpoint" was successful or not (and why).

In my case, I'm building a PWA.

**It can also run offline.**

When I try to fetch information via 'fetch' or other (into load function) if I can't take data, I ask the user to connect to fulfill the request or to proceed with other actions.
But how can I easily do it even with "page endpoints"?

I didn't find anything about it.
Direct me to some appropriate post if it can be done.

### Describe the proposed solution

**_The main solution is:_** 

Have a mechanism that prevents the presentation of an Internal Server Error (http status 500) in the event that the request 
to the "page endpoint" fails and that an exported function can be used to listen the request Response.

Ex.

export const catchPageEndpointResponse = async (response: Response): Promise&lt;void&gt; => {
      ....
};

The fact of exposing the function itself should prevent an Internal Server Error (http status 500)

_**Alternative 1:**_
have a mechanism that prevents the presentation of an Internal Server Error (http status 500) in the event 
that the request to the "page endpoint" generates a network timeout rather than an error.

In abstract:

Network or host Unreachable -> props = null (without 500)
Host reachable (data taken) with status 200 (or <400) -> props = {(any data)}
In all other cases, eg. 404, 500, etc. etc. -> an Internal Server Error (http status 500) as now.

_**Alternative 2:**_ (no good)
Totally avoid an Internal Server Error (http status 500)
props: Props (load function argument)

become:

server_response: Response (or something like that)

in order to have the complete answer provided by the "page endpoint".

### Alternatives considered

To overcome the problem I could very well use a simple endpoint and not a "page endpoint", but I would have to give up other possible advantages related to the use of "page endpoints" (maybe now they are not many, but in the future they could be many more).

It seems like a problem that can be underestimated, but generally if you don't force people to workaround it is always preferable.
I wish I would not be forced not to use a feature for a "trivially solvable" problem.
It is not a problem that can be overlooked precisely because it can be easily solved.

### Importance

would make my life easier

**Comment by Rich-Harris** [comment]
Network errors are errors — why wouldn't you want to show an error page? You can detect the fact that it was a network error either by duck-typing (unfortunately browsers aren't consistent here), or with `if (browser && !navigator.onLine)` inside the error page's `load` function.

If you want the app to continue to work offline, you can use a service worker.

**Comment by Rich-Harris** [comment]
I'm going to close this, as the current behaviour is deliberate. Errors are errors, and the default should be to show an error page when they are encountered. As you note, it's possible to do whatever you want within a `load` function, but that doesn't mean that the framework should encourage it.

**Comment by albaquarto** [comment]
I'm completly not agree with the definitive sentence of @Rich-Harris there are many case where navigate to an error page if an errore occurred is not the best thing to do. For example if the user compiled a very long and complex form and after for a temporary connection problem the request of submit fails the user don't want to recompile the form because the app navigate in an error page and the developer don't want implement a mechanism to restore the form. And even if the developer would implement the restore logic, the user must go back in the form page with an over click and this not is a good UX. in this case for example is most correct show an error toast message without navigate, so the user can retry to submit the form and without do to many click (remember that in a good UX only one more click is too much). Another similar case is when the backend  validation data fails and return an error text that suggest the modifications, please don't say me that the validation of form should be implemented in frontend because this is not always possible. I think that should be exist the possibility for the developer to implement a custom error handler that not navigate if is not necessary, with the possibility of call the "super" or default handler for restore the standard behavior if he want.

**Comment by giuseppe-palmeri** [comment]
It is true that errors are errors, but in this specific case there can be different types of errors that do not necessarily have to be treated as an absence of connection (no shadow endpoint => no page => Error Page)

If I'm not mistaken a shadow endpoint works like a common endpoint so it is the page that calls the endpoint and not vice versa.

For example I can have a page preloaded on the browser (PWA)
which goes to call the page endpoint (this was the case that occurred to me) and I did not want the flow to be redirected to an error page, but to warn the user of the momentary absence of line. The user could fill in the form and at the same time re-establish the line for sending the data.

Perhaps the problem lies in not being able to describe things properly. I hope I have explained myself better.

I believe it is fair to be able to gain more control over the shadow endpoint in specific or rare cases.

So I raise the question.

You don't have to redirect to the error page if you want to manually manage the error (Intercepting errors alone is more than enough for this).

The error page should not be shown if you go to expose
the function:

(ex) => {
... network error handling ...
}

So if I want more control over the shadow endpoint I can do it in rare cases like the one I encountered.

I hope I have explained myself better.
